src/docfx/Models/WatchCommand.cs

```diff
@@ -0,0 +1,179 @@
+                Serve(serveDirectory, settings.Host, settings.Port, settings.OpenBrowser, settings.OpenFile);
+            }
+            else if (settings.Watch)
+            {
+                using var watcher = Watch(baseDirectory, config.build, onChange);
+
+                // just block but here we can't use the host mecanism
+                // since we didn't start the server so use console one
+                using var canceller = new CancellationTokenSource();
+                Console.CancelKeyPress += (sender, args) => canceller.Cancel();
+                Task.Delay(Timeout.Infinite, canceller.Token).Wait();
+            }
+            else if (settings.Serve)
+            {
+                RunServe.Exec(serveDirectory, settings.Host, settings.Port, settings.OpenBrowser, settings.OpenFile);
+            }
+            else
+            {
+                onChange();
+            }
+        });
+    }
+
+    internal void Serve(string serveDirectory, string host, int? port, bool openBrowser, string openFile) {
+        if (CommandHelper.IsTcpPortAlreadyUsed(host, port))
+        {
+            Logger.LogError($"Serve option specified. But TCP port {port ?? 8080} is already being in use.");
+            return;
+        }
+        RunServe.Exec(serveDirectory, host, port, openBrowser, openFile);
+    }
+
+    // For now it is a simplistic implementation, in particular on the glob to filter mappping
+    // but it should be sufficient for most cases.
+    internal static IDisposable Watch(string baseDir, BuildJsonConfig config, Action onChange)
+    {
+        FileSystemWatcher watcher = new(baseDir)
+        {
+            IncludeSubdirectories = true,
+            NotifyFilter = NotifyFilters.Attributes | NotifyFilters.Size | NotifyFilters.FileName |
+                           NotifyFilters.DirectoryName | NotifyFilters.LastWrite
+        };
+
+        if (WatchAll(config))
+        {
+            watcher.Filters.Add("*.*");
+        }
+        else
+        {
+            RegisterFiles(watcher, config.Content);
+            RegisterFiles(watcher, config.Resource);
+
+            IEnumerable<string> forcedFiles = ["docfx.json", "*.md", "toc.yml"];
+            foreach (var forcedFile in forcedFiles)
+            {
+                if (!watcher.Filters.Any(f => f == forcedFile))
+                {
+                    watcher.Filters.Add(forcedFile);
+                }
+            }
+        }
+
+        // avoid to call onChange() in chain so await "last" event before re-rendering
+        var cancellation = new CancellationTokenSource[] { null };
+        async void debounce()
+        {
+            var token = new CancellationTokenSource();
+            lock (cancellation)
+            {
+                ResetToken(cancellation);
+                cancellation[0] = token;
+            }
+
+            await Task.Delay(100, token.Token);
+            if (!token.IsCancellationRequested)
+            {
+                onChange();
```

Comment:
if the build invoked by `onChange` method takes a minute to finish and files changes each second, would we trigger multiple build or wait for the last build to complete before triggering the next?